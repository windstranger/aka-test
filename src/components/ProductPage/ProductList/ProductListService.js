export const filterProductsBySelection = (productList, selectedFilters) => {
  return productList.filter(product => {
    const { colors, categories } = selectedFilters;
    let colorsRes = colors.findIndex(color => product.colors.includes(color.value));
    let hasCategory = categories.map(category => category.value).includes(product.category);
    return (!colors.length || colorsRes > -1) && (!categories.length || hasCategory);
  });
};