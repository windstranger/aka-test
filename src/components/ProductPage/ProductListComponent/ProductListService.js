export const filterProductsBySelection = (productList, selectedFilters) => {
  return productList.filter(product => {
    const { colors, categories } = selectedFilters;
    let colorsRes = colors.findIndex(color => product.colors.includes(color.value));
    let hasCategory = categories.map(category => category.value).includes(product.category);
    return (!colors.length || colorsRes > -1) && (!categories.length || hasCategory);
  });
};

export const getImageBySelectedColor = selectedProductFilters => productItem => {
  let colorIndex = -1;
  selectedProductFilters.colors.findIndex(color => {
    colorIndex = productItem.colors.indexOf(color.value);
    return colorIndex > -1;
  });

  return colorIndex > -1
    ? productItem.pictureToImage[colorIndex]
    : productItem.pictureToImage[0];
};
