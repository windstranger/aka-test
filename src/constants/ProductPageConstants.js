export const COLORS_LIST = {
  RED: 'red',
  YELLOW: 'yellow',
  WHITE: 'white',
  ORANGE: 'orange',
  GREEN: 'green',
  BLUE: 'blue',
  VIOLET: 'violet',
  BLACK: 'black',
  PINK: 'pink'
};

export const CATEGORIES_LIST = {
  BAGS: 'bags',
  BACKPACKS: 'backpacks',
  ACCESSORIES: 'accessories',
  LUGGAGE: 'luggage'
};
