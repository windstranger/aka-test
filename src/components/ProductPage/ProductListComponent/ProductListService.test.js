import { CATEGORIES_LIST, COLORS_LIST } from '../../../constants/ProductPageConstants';
import { filterProductsBySelection, getImageBySelectedColor } from './ProductListService';

let productList = [
  {
    id: 1,
    colors: [
      COLORS_LIST.BLACK,
      COLORS_LIST.RED,
      COLORS_LIST.BLUE,
      COLORS_LIST.PINK
    ],
    category: CATEGORIES_LIST.ACCESSORIES
  },
  {
    id: 2,
    colors: [
      COLORS_LIST.WHITE,
      COLORS_LIST.GREEN,
      COLORS_LIST.ORANGE,
      COLORS_LIST.PINK
    ],

    category: CATEGORIES_LIST.BACKPACKS
  },
  {
    id: 3,
    colors: [
      COLORS_LIST.VIOLET,
      COLORS_LIST.YELLOW,
      COLORS_LIST.BLACK,
      COLORS_LIST.PINK
    ],
    category: CATEGORIES_LIST.BACKPACKS
  },
  {
    id: 4,
    colors: [
      COLORS_LIST.BLACK,
      COLORS_LIST.ORANGE,
      COLORS_LIST.YELLOW,
      COLORS_LIST.PINK
    ],
    category: CATEGORIES_LIST.LUGGAGE
  }
];

test('should get image on color selection', () => {
  let productItem = {
    pictureToImage: [1, 2],
    colors: ['red', 'yellow']
  };
  expect(getImageBySelectedColor({ colors: [{ value: 'red' }] })(productItem) === 1);
  expect(getImageBySelectedColor({ colors: [{ value: 'yellow' }] })(productItem) === 2);
  expect(getImageBySelectedColor({ colors: [] })(productItem) === 1);
});

test('should filter productList on color', () => {
  let selectedFilters = { colors: [{ value: 'red' }], categories: [] };
  expect(filterProductsBySelection(productList, selectedFilters)).toMatchSnapshot();
  selectedFilters = { colors: [{ value: 'red' }, { value: 'yellow' }], categories: [] };
  expect(filterProductsBySelection(productList, selectedFilters)).toMatchSnapshot();
});

test('should filter productList on category', () => {
  let selectedFilters = { colors: [], categories: [{ value: CATEGORIES_LIST.LUGGAGE }] };
  expect(filterProductsBySelection(productList, selectedFilters)).toMatchSnapshot();
  selectedFilters = { colors: [], categories: [{ value: CATEGORIES_LIST.BACKPACKS }] };
  expect(filterProductsBySelection(productList, selectedFilters)).toMatchSnapshot();
});

test('should filter productList on category and color', () => {
  let selectedFilters = {
    colors: [{ value: COLORS_LIST.RED }, { value: COLORS_LIST.YELLOW }],
    categories: [{ value: CATEGORIES_LIST.LUGGAGE }]
  };
  expect(filterProductsBySelection(productList, selectedFilters)).toMatchSnapshot();
  selectedFilters = {
    colors: [],
    categories: []
  };
  expect(filterProductsBySelection(productList, selectedFilters)).toMatchSnapshot();
});
