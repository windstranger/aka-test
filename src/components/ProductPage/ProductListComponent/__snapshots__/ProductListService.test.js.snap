// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should filter productList on category 1`] = `
Array [
  Object {
    "category": "luggage",
    "colors": Array [
      "black",
      "orange",
      "yellow",
      "pink",
    ],
    "id": 4,
  },
]
`;

exports[`should filter productList on category 2`] = `
Array [
  Object {
    "category": "backpacks",
    "colors": Array [
      "white",
      "green",
      "orange",
      "pink",
    ],
    "id": 2,
  },
  Object {
    "category": "backpacks",
    "colors": Array [
      "violet",
      "yellow",
      "black",
      "pink",
    ],
    "id": 3,
  },
]
`;

exports[`should filter productList on category and color 1`] = `
Array [
  Object {
    "category": "luggage",
    "colors": Array [
      "black",
      "orange",
      "yellow",
      "pink",
    ],
    "id": 4,
  },
]
`;

exports[`should filter productList on category and color 2`] = `
Array [
  Object {
    "category": "accessories",
    "colors": Array [
      "black",
      "red",
      "blue",
      "pink",
    ],
    "id": 1,
  },
  Object {
    "category": "backpacks",
    "colors": Array [
      "white",
      "green",
      "orange",
      "pink",
    ],
    "id": 2,
  },
  Object {
    "category": "backpacks",
    "colors": Array [
      "violet",
      "yellow",
      "black",
      "pink",
    ],
    "id": 3,
  },
  Object {
    "category": "luggage",
    "colors": Array [
      "black",
      "orange",
      "yellow",
      "pink",
    ],
    "id": 4,
  },
]
`;

exports[`should filter productList on color 1`] = `
Array [
  Object {
    "category": "accessories",
    "colors": Array [
      "black",
      "red",
      "blue",
      "pink",
    ],
    "id": 1,
  },
]
`;

exports[`should filter productList on color 2`] = `
Array [
  Object {
    "category": "accessories",
    "colors": Array [
      "black",
      "red",
      "blue",
      "pink",
    ],
    "id": 1,
  },
  Object {
    "category": "backpacks",
    "colors": Array [
      "violet",
      "yellow",
      "black",
      "pink",
    ],
    "id": 3,
  },
  Object {
    "category": "luggage",
    "colors": Array [
      "black",
      "orange",
      "yellow",
      "pink",
    ],
    "id": 4,
  },
]
`;
